File: main/client/src/mill/main/client/lock/Locks.java
package mill.main.client.lock;

import mill.main.client.ServerFiles;

public final class Locks implements AutoCloseable {

  public final Lock clientLock;
  public final Lock processLock;

  public Locks(Lock clientLock, Lock processLock) {
    this.clientLock = clientLock;
    this.processLock = processLock;
  }

  public static Locks files(String serverDir) throws Exception {
    return new Locks(
        new FileLock(serverDir + "/" + ServerFiles.clientLock),
        new FileLock(serverDir + "/" + ServerFiles.processLock));
  }

  public static Locks memory() {
    return new Locks(new MemoryLock(), new MemoryLock());
  }

  @Override
  public void close() throws Exception {
    clientLock.delete();
    processLock.delete();
  }
}

File: main/client/src/mill/main/client/lock/Locks.scala
package mill.main.client.lock

import mill.main.client.ServerFiles

final case class Locks(val clientLock: Lock, val processLock: Lock) extends AutoCloseable {

  @throws[Exception]
  override def close(): Unit = {
    clientLock.delete()
    processLock.delete()
  }
}

object Locks {
  @throws[Exception]
  def files(serverDir: String): Locks = {
    new Locks(
      new FileLock(s"$serverDir/${ServerFiles.clientLock}"),
      new FileLock(s"$serverDir/${ServerFiles.processLock}")
    )
  }

  def memory(): Locks = {
    new Locks(new MemoryLock(), new MemoryLock())
  }
}
